monitor/src/monitor.rs

```
@@ -0,0 +1,276 @@
+                    self.root_contexts.retain(|(instant, context)| {
+                        if instant.elapsed() > self.event_window {
+                            println!("{}", serde_json::to_string_pretty(context).unwrap());
+                            return false;
+                        }
+                        true
+                    });
+                },
+                _ = shutdown_receiver.recv() => break,
+            }
+        }
+
+        Ok(())
+    }
+}
+
+async fn shutdown(
+    shutdown_receiver: &mut broadcast::Receiver<()>,
+    shutdown_sender: &broadcast::Sender<()>,
+) -> Result<()> {
+    tokio::select! {
+        maybe_value = signal::ctrl_c() => {
+            if let Err(e) = maybe_value {
+                info!(error = %e, "error receiving ctrl-c")
+            }
+            info!("shutting down event broker");
+            if let Err(e) = shutdown_sender.send(()) {
+                info!(error = %e, "unable to send shutdown");
+            }
+        },
+        _ = shutdown_receiver.recv() => (),
+    }
+    Ok(())
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    let config = config::Config::new()?;
+
+    tracing_subscriber::registry()
+        .with(fmt::layer())
+        .with(EnvFilter::from_default_env())
+        .try_init()?;
+
+    let mut reader = Reader::new(&config.log_file)?;
+    let mut publisher = Publisher::new(config.event_window, &config.scope);
+
+    let (event_tx, event_rx) = mpsc::channel::<EventGroup>(10);
+    let mut event_rx = ReceiverStream::new(event_rx);
+
+    let (shutdown_tx, mut shutdown_rx1) = broadcast::channel::<()>(2);
+    let mut shutdown_rx2 = shutdown_tx.subscribe();
+    let mut shutdown_rx3 = shutdown_tx.subscribe();
+
+    try_join!(
+        shutdown(&mut shutdown_rx1, &shutdown_tx),
+        reader.read(&event_tx, &mut shutdown_rx2),
+        publisher.publish(&mut event_rx, &mut shutdown_rx3),
+    )
+    .map(|_| ())
```

doesn't that swallow errors returned from the futures?